BUILD: Output gadget files

mooeypoo opened the pull request:
*NOTE*: This is written on top of #8 and should only be merged after that PR is merged

- Add a build step to output gadget files on top of browser extension files.
- Move generated output into `dist/` folder for cleanliness

mooeypoo committed:
Convert to ES6 and add mocha tests and babel to build steps

- Change the code syntax to ES6
- Adjust tests to use mocha and karma with a polyfill for `document`
- Adjust test syntax to ES6
- Adjust the build steps to use browserify with babel
- Use `src/app.js` as an entrypoint for the runnable code

mooeypoo committed:
BUILD: Output gadget files

- Create a build step for the gadget file
- Add a gadget template to output the code into
- Add a specific gadget banner and template
- Reorganize the output into `dist` folder that contains
  both outputs:
  - `dist/extension` for the browser extension
  - `dist/gadget` for the gadget files

samwilson commented on build/header.gadget.txt:
```diff
@@ -0,0 +1,18 @@
+ * Any updates should be created from the official repository
+ * by running `grunt build` and using
+ * `dist/gadget/generated.WhoWroteThat.gadget.js` and
+ * `dist/gadget/generated.WhoWroteThat.gadget.css` and
```
These paths are different to what's elsewhere in this patch, I think.

mooeypoo replied:
Whoops, that's right, I'm fixing.

mooeypoo commented:
Comments fixed, I squashed the commits into one with a more comprehensive commit message (copied into the PR)

Ready for review; ONLY MERGE THIS AFTER THE ES6 PR! :)

cimurah committed:
Merge pull request #8 from wikimedia/convert-to-es6

Convert to ES6 and add mocha tests and babel to build steps

mooeypoo committed:
Fix typo in parameter documentation for Api.js

samwilson committed:
Merge pull request #14 from wikimedia/fix-typo-docs

Fix typo in parameter documentation for Api.js

mooeypoo committed:
BUILD: Add i18n file support

- Add a `i18n` folder with language json files
- Add a Grunt job to read available language files into a json object
- Add async grunt job to produce the final file; this job has to make
  sure that browserify generates the output before we can run 'replace'
  to replace the values in the template and produce the final production
  file.
- Add banana checker to the i18n files
- Inject the object into the generated script
- Load messages throuw `mw.messages.set`

samwilson committed:
Merge pull request #15 from wikimedia/add-lang-file-support

BUILD: Add i18n file support

samwilson commented on Gruntfile.js:
```diff
@@ -49,67 +77,49 @@ module.exports = function Gruntfile( grunt ) {
-							src +
-							'/* >> End source: ' + filepath + ' << */';
-					}
+					banner: grunt.file.read( 'build/header.gadget.txt' )
```
Should this header also be added to the generated JS file?

mooeypoo replied:
So, I thought about that, but the generated JS is inserted *into* the template.

I was thinking of putting it up at the top of the template, but I need to look into how to make it replace the variables if I do that, so I figured we can do that as maybe a followup.

samwilson replied:
Sure, let's leave it for after. This is all working in the important ways.

samwilson reviewed (approved):
Works in my testing! :)
